Core/MIPS/IR/IRPassSimplify.cpp

```diff
@@ -0,0 +1,199 @@
+#include "Core/MIPS/IR/IRPassSimplify.h"
+#include "Core/MIPS/IR/IRRegCache.h"
+
+void SimplifyInPlace(IRInst *inst, int count, const u32 *constPool) {
+	for (int i = 0; i < count; i++) {
+		switch (inst[i].op) {
+		case IROp::AddConst:
+			if (constPool[inst[i].src2] == 0)
+				inst[i].op = IROp::Mov;
+			else if (inst[i].src1 == 0) {
+				inst[i].op = IROp::SetConst;
+				inst[i].src1 = inst[i].src2;
+			}
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+
+u32 Evaluate(u32 a, u32 b, IROp op) {
+	switch (op) {
+	case IROp::Add: case IROp::AddConst: return a + b;
+	case IROp::Sub: case IROp::SubConst: return a - b;
+	case IROp::And: case IROp::AndConst: return a & b;
+	case IROp::Or: case IROp::OrConst: return a | b;
+	case IROp::Xor: case IROp::XorConst: return a ^ b;
+	case IROp::Shr: case IROp::ShrImm: return a >> b;
+	case IROp::Sar: case IROp::SarImm: return (s32)a >> b;
+	case IROp::Ror: case IROp::RorImm: return (a >> b) | (a << (32 - b));
+	case IROp::Shl: case IROp::ShlImm: return a << b;
+	case IROp::Slt: case IROp::SltConst: return ((s32)a < (s32)b);
+	case IROp::SltU: case IROp::SltUConst: return (a < b);
+	default:
+		return -1;
+	}
+}
+
+IROp ArithToArithConst(IROp op) {
+	switch (op) {
+	case IROp::Add: return IROp::AddConst;
+	case IROp::Sub: return IROp::SubConst;
+	case IROp::And: return IROp::AndConst;
+	case IROp::Or: return IROp::OrConst;
+	case IROp::Xor: return IROp::XorConst;
+	case IROp::Slt: return IROp::SltConst;
+	case IROp::SltU: return IROp::SltUConst;
+	default:
+		return (IROp)-1;
+	}
+}
+
+
+bool PropagateConstants(const IRWriter &in, IRWriter &out) {
+	IRRegCache gpr(&out);
+
+	const u32 *constants = in.GetConstants().data();
+	bool logBlocks = false;
+	for (int i = 0; i < (int)in.GetInstructions().size(); i++) {
+		IRInst inst = in.GetInstructions()[i];
+		bool symmetric = true;
+		switch (inst.op) {
```

Comment:
We should do something like `if (out.GetContants.size() > 128) goto doDefault;`...

-[Unknown]